### PR TITLE
Cm/log tensorboard

### PR DESCRIPTION
**Pull request by chMoussa**
Related to Issue #444 : 

- Fixing tensorboard logs to always include the loss/metrics values at iteration 0 (evaluating models before training) and fix inconsistencies wrt iteration numbers when logging training and validation values.
- Avoiding logging math.nan in tensorboard for the training loss.

### CODE REVIEW

File: qadence/ml_tools/train_grad.py [hunk 1]
```diff
@@ -154,12 +154,32 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+                best_val_loss, metrics = loss_fn(model, xs_to_device)
+
+                metrics["val_loss"] = best_val_loss
+                write_tensorboard(writer, math.nan, metrics, init_iter)
+
+            if config.folder:
+                if config.checkpoint_best_only:
+                    write_checkpoint(config.folder, model, optimizer, iteration="best")
+                else:
+                    write_checkpoint(config.folder, model, optimizer, init_iter)
```
Comment (DanieleCucurachi) [inline qadence/ml_tools/train_grad.py:176]: This indeed includes the pre-training loss value among the validation losses in the tensorboard logs.

File: qadence/ml_tools/train_grad.py [hunk 2]
```diff
@@ -154,12 +154,32 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+
+        except KeyboardInterrupt:
+            logger.info("Terminating training gracefully after the current iteration.")
 
         # outer epoch loop
+        init_iter += 1
```
Comment (DanieleCucurachi) [inline qadence/ml_tools/train_grad.py:182]: Changing the iteration value resolves the issue of the final row of tensorboard logged losses being doubled (you can observe this in the current version of `qadence` by printing out the tensorboard logs). This problem was not outlined in the original issue description as it was discovered later, however this change fixes it so no need to add it now.

File: qadence/ml_tools/train_grad.py [hunk 3]
```diff
@@ -217,8 +240,19 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
             except KeyboardInterrupt:
                 logger.info("Terminating training gracefully after the current iteration.")
                 break
+        try:
+            xs = next(dl_iter) if dataloader is not None else None  # type: ignore[arg-type]
+            xs_to_device = data_to_device(xs, device=device, dtype=data_dtype)
+            loss, metrics = loss_fn(model, xs_to_device)
+            if "val_loss" in metrics:
+                del metrics["val_loss"]
```
Comment (DanieleCucurachi) [inline qadence/ml_tools/train_grad.py:248]: @chMoussa could you please explain why the `val_loss` key-value pair is removed in this section and why the validation loss is being evaluated on the training dataset? I am not sure I understand the purpose of this.

File: qadence/ml_tools/train_grad.py [hunk 4]
```diff
@@ -217,8 +240,19 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+                print_metrics(loss, metrics, iteration)
+
+        except KeyboardInterrupt:
+            logger.info("Terminating training gracefully after the current iteration.")
 
-    # Final writing and checkpointing
+    # Final printing, writing and checkpointing
     if config.folder and not config.checkpoint_best_only:
         write_checkpoint(config.folder, model, optimizer, iteration)
     write_tensorboard(writer, loss, metrics, iteration)
```
Comment (DanieleCucurachi) [inline qadence/ml_tools/train_grad.py:258]: Currently, the final loss is logged by default regardless of `config.write_every`. Should we consider logging the final validation loss by default as well?

File: qadence/ml_tools/printing.py [hunk 5]
```diff
@@ -12,7 +14,8 @@ def print_metrics(loss: float | None, metrics: dict, iteration: int) -> None:
 
 
 def write_tensorboard(writer: SummaryWriter, loss: float, metrics: dict, iteration: int) -> None:
-    writer.add_scalar("loss", loss, iteration)
+    if not isnan(loss):
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/ml_tools/printing.py:17]: @DanieleCucurachi good remarks! I like the solution proposed for 1. Will get on it!

**Review by DanieleCucurachi — APPROVED** [review]
Thank you @chMoussa, the new `write_tensorboard()` implementation is definitely better. Approved!

**Comment by chMoussa** [comment]
Closes issue #444.
